### PR TITLE
Implemented all existing datapar algorithms using Boost.SIMD

### PR DESCRIPTION
**Pull request by hkaiser**
- factoring out code common to both backends
- adding build system support for Boost.SIMD
- unified configurations for Vc and Boost.SIMD
- flyby: cleaning up the Vc backend

### CODE REVIEW

File: cmake/HPX_SetupBoostSIMD.cmake [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  hpx_error("Using Boost.SIMD requires to set the variable BOOST_SIMD_ROOT to the Boost.SIMD installation directory.")
+endif()
+if(Boost_VERSION LESS 106000)
+  hpx_error("Boost.SIMD requires using Boost V1.60 or above")
+endif()
+
+include_directories(SYSTEM ${BOOST_SIMD_ROOT}/include)
+
+hpx_add_config_define(HPX_HAVE_DATAPAR)
+hpx_add_config_define(HPX_HAVE_DATAPAR_BOOST_SIMD)
```
Comment (sithhell) [inline cmake/HPX_SetupBoostSIMD.cmake]: What about special compile flags? How will the intrinsics get enabled for the compiler (usually require special compiler switches)?

File: hpx/parallel/datapar/transform_loop.hpp [hunk 2]
```diff
@@ -9,82 +8,18 @@
...
 
-#if defined(HPX_HAVE_VC_DATAPAR)
-#include <hpx/util/decay.hpp>
-#include <hpx/util/invoke.hpp>
-#include <hpx/util/tuple.hpp>
-
-#include <hpx/parallel/datapar/detail/iterator_helpers.hpp>
+#if defined(HPX_HAVE_DATAPAR)
+#include <hpx/parallel/datapar/detail/transform_loop_vc.hpp>
+#include <hpx/parallel/datapar/detail/transform_loop_boost_simd.hpp>
```
Comment (sithhell) [inline hpx/parallel/datapar/transform_loop.hpp]: shouldn't this include just one of those?

File: cmake/HPX_SetupBoostSIMD.cmake (hunk 1 above)
Comment (hkaiser) [inline cmake/HPX_SetupBoostSIMD.cmake]: That is something Joel is still working on. He said they would provide a cmake macro for this information. For now this is all we can do.

Comment (hkaiser) [inline cmake/HPX_SetupBoostSIMD.cmake]: @jfalcou: any news on this by any chance?

**Comment by hkaiser** [comment]
@sithhell can this go ahead now?

**Comment by hkaiser** [comment]
@sithhell Can we merge this now? I really would like to get this in to be able to continue working on unifying the backend API.

**Comment by sithhell** [comment]
What happens when we include the Boost.SIMD datapar component without any additional compile flags? Will it still compile? What's the effect then?

**Comment by hkaiser** [comment]
> What happens when we include the Boost.SIMD datapar component without any additional compile flags? Will it still compile? What's the effect then?

Good question. Why don't you tell me? It works out of the box for MSVC.

**Comment by sithhell** [comment]
I don't know enough about Boost.SIMD in order to know what happens when compiled without, for example, -mavx for gcc. Looking at the code, it seems that it just defaults to scalar code if the different SIMD flags aren't enabled. @jfalcou, can you clarify?

**Comment by jfalcou** [comment]
The real culprit is that each OS/compiler have a different default behavior : 

- MSVC in 32 bits doesn't use any SIMD
- MSVC in 64 bits use SSE2 by default without needing any option to be set
- gcc/clang generates scalar code on 32bits LINUX
- gcc/clang generates SSE2 automatically on 64 bits LINUX

I would advice to use -march=native on gcc/clang by default so it selects the proper arch model and activates the proper SIMD level for the architecture (including SSE>2 and AVX).

On MSVC, you need to put the proper /arch: option to get the proper level.

**Comment by sithhell** [comment]
I guess having it work (compile and run) with the default settings is good enough. Users can always tune the used flags as they like.

**Pull request merged**